### PR TITLE
Update URLEncodedFormEncoder encoding rules

### PR DESCRIPTION
**Pull request by ptoffy**
Solves #3173 
References https://url.spec.whatwg.org/#application-x-www-form-urlencoded-percent-encode-set to set the encoding rules 
> The [application/x-www-form-urlencoded percent-encode set](https://url.spec.whatwg.org/#application-x-www-form-urlencoded-percent-encode-set) contains all code points, except the [ASCII alphanumeric](https://infra.spec.whatwg.org/#ascii-alphanumeric), U+002A (*), U+002D (-), U+002E (.), and U+005F (_).

### CODE REVIEW

File: Sources/Vapor/URLEncodedForm/URLEncodedFormSerializer.swift [hunk 1]
```diff
@@ -75,7 +75,7 @@ private enum Characters {
     static let allowedCharacters: CharacterSet = {
         var allowed = CharacterSet.urlQueryAllowed
         // these symbols are reserved for url-encoded form
-        allowed.remove(charactersIn: "?&=[];+")
+        allowed.remove(charactersIn: "?&=[];+$")
```
Comment (0xTim) [inline Sources/Vapor/URLEncodedForm/URLEncodedFormSerializer.swift:78]: We should inverse this completely as per https://url.spec.whatwg.org/#application-x-www-form-urlencoded-percent-encode-set.

Essentially if it _isn't_ alphanumeric, `*`, `-`, `.` or `_` we should encode it

**Review by 0xTim — CHANGES_REQUESTED** [review]

File: Sources/Vapor/URLEncodedForm/URLEncodedFormSerializer.swift [hunk 2]
```diff
@@ -72,10 +72,10 @@ extension String {
 
 /// Characters allowed in form-urlencoded data.
 private enum Characters {
+    // https://url.spec.whatwg.org/#application-x-www-form-urlencoded-percent-encode-set
     static let allowedCharacters: CharacterSet = {
-        var allowed = CharacterSet.urlQueryAllowed
-        // these symbols are reserved for url-encoded form
-        allowed.remove(charactersIn: "?&=[];+")
+        var allowed = CharacterSet.alphanumerics
```
Comment (ptoffy) [inline Sources/Vapor/URLEncodedForm/URLEncodedFormSerializer.swift:77]: I'm honestly not sure why this is allowing `[` and `]`

File: Sources/Vapor/URLEncodedForm/URLEncodedFormSerializer.swift [hunk 3]
```diff
@@ -72,10 +72,10 @@ extension String {
...
+        allowed.insert(charactersIn: "*-._")
```
Comment (ptoffy) [inline Sources/Vapor/URLEncodedForm/URLEncodedFormSerializer.swift:78]: I'm honestly not sure why this is allowing `[` and `]`

**Review by 0xTim — APPROVED** [review]

File: Tests/VaporTests/URLEncodedFormTests.swift [hunk 4]
```diff
@@ -157,65 +157,63 @@ final class URLEncodedFormTests: XCTestCase {
...
-                Date(timeIntervalSince1970: 10_000),
-                Date(timeIntervalSince1970: 20_000),
-                Date(timeIntervalSince1970: 30_000),
-                Date(timeIntervalSince1970: 40_000),
-                Date(timeIntervalSince1970: 50_000),
+                Date(timeIntervalSince1970: 10000),
+                Date(timeIntervalSince1970: 20000),
+                Date(timeIntervalSince1970: 30000),
+                Date(timeIntervalSince1970: 40000),
+                Date(timeIntervalSince1970: 50000),
```
Comment (gwynne) [inline Tests/VaporTests/URLEncodedFormTests.swift:169]: Honestly, turn down the settings on your code formatter, this is a change for the worse 🤣

**Review by gwynne — APPROVED** [review]

**Pull request merged**